### PR TITLE
`selector-pseudo-class-parentheses-space-inside` rule

### PR DESCRIPTION
**Pull request by evilebottnawi**
https://github.com/stylelint/stylelint/issues/1073

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@jeddy3 review needed :smile:

File: src/rules/selector-pseudo-class-parentheses-space-inside/index.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      ],
+    })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      selectorParser(selectorTree => {
+        selectorTree.eachPseudo(pseudoNode => {
+          if (pseudoNode.parent
+            && pseudoNode.parent.parent
+            && pseudoNode.parent.parent.type === "pseudo"
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-parentheses-space-inside/index.js]: Can be simplified with `_.get()`: `if (_.get(pseudoNode, 'parent.parent.type) === "pseudo")`

**Comment by davidtheclark** [comment]
LGTM. I just had one unimportant suggestion.

File: src/rules/selector-pseudo-class-parentheses-space-inside/index.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          ) { return }
+
+          const pseudoSelectorString = pseudoNode.toString()
+
+          styleSearch({ source: pseudoSelectorString, target: "(" }, match => {
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-parentheses-space-inside/index.js]: Are you sure that there is not already data about the spacing within the postcss-selector-parser node? I suspect that there is, right?

Comment (evilebottnawi) [inline src/rules/selector-pseudo-class-parentheses-space-inside/index.js]: @davidtheclark no, in `spaces`: `before` and `after` we have spaces:
```css
a :not ()
  ↑     ↑
These spaces
```

Comment (evilebottnawi) [inline src/rules/selector-pseudo-class-parentheses-space-inside/index.js]: not right `spaces` always for `pseudo` are empty. We don't have information about spaces in `pseudoNode`.

File: docs/user-guide/rules.md [hunk 3]
```diff
@@ -142,6 +142,7 @@ Here are all the rules within stylelint, grouped by the [*thing*](http://apps.wo
 - [`selector-no-universal`](../../src/rules/selector-no-universal/README.md): Disallow the universal selector.
 - [`selector-no-vendor-prefix`](../../src/rules/selector-no-vendor-prefix/README.md): Disallow vendor prefixes for selectors.
 - [`selector-pseudo-class-case`](../../src/rules/selector-pseudo-class-case/README.md): Specify lowercase or uppercase for pseudo-class selectors.
+- [`selector-pseudo-class-parentheses-space-inside`](../../src/rules/selector-parentheses-space-inside/README.md): Require a single space or disallow whitespace on the inside of the parentheses within selectors.
```
Comment (jeddy3) [inline docs/user-guide/rules.md]: "Require a single space or disallow whitespace on the inside of the parentheses within pseudo-class selectors."

File: src/rules/selector-pseudo-class-parentheses-space-inside/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# selector-pseudo-class-parentheses-space-inside
+
+Require a single space or disallow whitespace on the inside of the parentheses within selectors.
```
Comment (jeddy3) [inline src/rules/selector-pseudo-class-parentheses-space-inside/README.md]: "Require a single space or disallow whitespace on the inside of the parentheses within pseudo-class selectors."

File: src/rules/selector-pseudo-class-parentheses-space-inside/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+```css
+input:not( [type="submit"] ) {}
+/**       ↑               ↑
```
Comment (jeddy3) [inline src/rules/selector-pseudo-class-parentheses-space-inside/README.md]: Can we align the arrows with the parens please?

File: src/rules/selector-pseudo-class-parentheses-space-inside/index.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      selectorParser(selectorTree => {
+        selectorTree.eachPseudo(pseudoNode => {
+          if (_.get(pseudoNode, "parent.parent.type") === "pseudo") { return }
+
+          const pseudoSelectorString = pseudoNode.toString()
+
+          styleSearch({ source: pseudoSelectorString, target: "(" }, match => {
+            const nextCharIsSpace = pseudoSelectorString[match.startIndex + 1] === " "
+            if (nextCharIsSpace && expectation === "never") {
+              report({
```
Comment (jeddy3) [inline src/rules/selector-pseudo-class-parentheses-space-inside/index.js]: Can we minimise the duplication within `report()` by using our idiomatic `complain()` pattern as [seen here](https://github.com/stylelint/stylelint/blob/master/src/rules/comment-whitespace-inside/index.js#L45-L66)?

**Comment by jeddy3** [comment]
Looking great! I've made a few requests though.

File: src/rules/selector-pseudo-class-parentheses-space-inside/index.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      actual: expectation,
+      possible: [
+        "always",
+        "never",
+      ],
+    })
+    if (!validOptions) { return }
+
+    root.walkRules(rule => {
+      selectorParser(selectorTree => {
```
Comment (davidtheclark) [inline src/rules/selector-pseudo-class-parentheses-space-inside/index.js]: Let's also do an optimization here, by not parsing the selector if there is no `(` within it.

**Comment by evilebottnawi** [comment]
@davidtheclark @jeddy3 ping

**Comment by jeddy3** [comment]
Thanks for making the changes! LGTM.

**Pull request merged**